Implement LMDB automatic resizing

delta1 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
This commit implements the functionality to automatically resize LMDB environment mapsize as it fills up. This should ensure that new data can always be inserted as long as there is disk space available. 

There were a few minor issues fixed: spelling, docs, and removing an unused `use` declaration. 

Thanks to @hansieodendaal for his contributions

## Motivation and Context
Issue #2181 

## How Has This Been Tested?
* [X] Existing tests pass 
* [X] New tests created 
* [X] Tested manually on macOS 
* [ ] Testing on Windows/Linux still required

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [X] Bug fix (non-breaking change which fixes an issue)
* [X] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [X] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [X] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [X] I'm merging against the `development` branch.
* [X] I ran `cargo-fmt --all` before pushing.
* [X] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [X] I have added tests to cover my changes.

hansieodendaal commented on infrastructure/storage/src/lmdb_store/store.rs:
```diff
@@ -359,11 +368,61 @@ impl LMDBStore {
+    }
+
+    /// Returns a result tuple indicating if the LMDB Environment should be resized, along with the new size in bytes.
+    /// Currently uses a contant resize threshold.
```
```suggestion
    /// Currently uses a constant resize threshold.
```

delta1 replied:
Thanks Hansie nice catch - fixed

hansieodendaal commented on infrastructure/storage/src/lmdb_store/store.rs:
```diff
@@ -359,11 +368,61 @@ impl LMDBStore {
+        let env_info = env.info()?;
+        let psize = env.stat()?.psize as usize;
+        let size_used_bytes = psize * env_info.last_pgno;
+        let new_size = size_used_bytes + init_size_bytes;
+        let needs_resize = should_resize(size_used_bytes, env_info.mapsize, RESIZE_THRESHOLD);
+
+        trace!(
+            target: LOG_TARGET,
+            "LMDB mapsize: {}, bytes used: {}, bytes remaining: {}, used: {:.*}, threshold: {}, resize required: {}.",
+            env_info.mapsize,
+            size_used_bytes,
+            env_info.mapsize - size_used_bytes,
+            4, // decimal places for "used"
+            size_used_bytes as f64 / env_info.mapsize as f64,
+            RESIZE_THRESHOLD,
+            needs_resize.to_string()
+        );
```
```suggestion
        let needs_resize = should_resize(size_used_bytes, env_info.mapsize, RESIZE_THRESHOLD);

        let new_size = match needs_resize {
            true => {
                debug!(
                    target: LOG_TARGET,
                    "LMDB mapsize resize required: {}, bytes used: {}, bytes remaining: {}, used: {:.*}, threshold: \
                     {}.",
                    env_info.mapsize,
                    size_used_bytes,
                    env_info.mapsize - size_used_bytes,
                    4, // decimal places for "used"
                    size_used_bytes as f64 / env_info.mapsize as f64,
                    RESIZE_THRESHOLD,
                );
                size_used_bytes + init_size_bytes
            },
            false => init_size_bytes,
        };
```

delta1 replied:
Thanks for the suggestion Hansie - I'll certainly take a look at this. It shouldn't return `init_size_bytes` though - as the env size may have already grown beyond that and so this would actually try to decrease the mapsize.

hansieodendaal commented on infrastructure/storage/tests/lmdb.rs:
```diff
@@ -330,3 +330,47 @@ fn lmbd_resize_on_create() {
+
+            // check the allocated size is now greater than it was initially
+            assert!(page_size_total > 1024 * 1024);
+            assert!(percent_left > 0.0);
```
I think in this instance we should also assert that `space left <= (1 + RESIZE_THRESHOLD ) * initial capacity`, to ensure it does not grow out of bounds.

delta1 replied:
Thanks I'll take a look at that

hansieodendaal replied:
I will post new suggestions; ran a 'sync from scratch' test and the resize logic has some problem.

hansieodendaal reviewed (commented):
Nice one! 

- My one concern is that I think it will be overkill to call the trace log in `pub fn resize_required` for every db write operation, which will happen as currently implemented. See proposed.
- Also added a test suggestion to check that the db does not grow out of bounds.

Regards

hansieodendaal reviewed (commented):
See https://github.com/delta1/tari/pull/1 for updated review suggestions; I have redone my previous suggestions.

delta1 commented:
Thanks @hansieodendaal - I'll merge that and test it out! Appreciate your assistance

hansieodendaal commented on infrastructure/storage/src/lmdb_store/store.rs:
```diff
@@ -359,11 +352,40 @@ impl LMDBStore {
+        let env_info = env.info()?;
+        let size_used_bytes = env.stat()?.psize as usize * env_info.last_pgno;
+        let needs_resize = env_info.mapsize - size_used_bytes <= (init_size_bytes as f64 * RESIZE_THRESHOLD) as usize;
+        let path = env.path().unwrap().to_str().unwrap();
```
```suggestion
```

hansieodendaal replied:
Realized this is an unnecessary call each time, could be simplified

hansieodendaal commented on infrastructure/storage/src/lmdb_store/store.rs:
```diff
@@ -359,11 +352,40 @@ impl LMDBStore {
+            debug!(
+                target: LOG_TARGET,
+                "({}) LMDB size used {:?} MB, environment space left {:?} MB, increased by {:?} MB.",
+                path,
```
```suggestion
                env.path().unwrap().to_str().unwrap(),
```

hansieodendaal replied:
As per above

stringhandler commented on base_layer/core/src/chain_storage/error.rs:
```diff
@@ -85,6 +86,11 @@ pub enum ChainStorageError {
     #[error("Value not found: {0}")]
     LmdbValueNotFound(lmdb_zero::Error),
+    #[error("LMDB error: {source}")]
+    LMDBError {
```
```suggestion
    LmdbError {
```

stringhandler commented on base_layer/core/src/chain_storage/lmdb_db/lmdb.rs:
```diff
@@ -46,9 +46,9 @@ where T: Serialize {
     bincode::serialize_into(&mut buf, data)
-        .or_else(|e| {
+        .map_err(|e| {
             error!(target: LOG_TARGET, "Could not serialize lmdb: {:?}", e);
```
nit: Could probably just move the log into the `map_err()` below, instead of having 2 `map_err`s

delta1 replied:
Good idea - will fix

stringhandler reviewed (commented):
One or two nits, but overall happy with how you've done it. I think that RESIZE_THRESHOLD should be moved to the config, with a default value as well as initial size and a new variable for size to grow by. This allows users to control disk usage. I think that RESIZE_THRESHOLD should also be a number in MB and not a percentage. The reason for that is the DB is always growing by a constant size which is `max_block_size * block_time`, so we could use the resize threshold to always add enough space for `x` number of blocks, rather than keeping 50% for growth.
I suggest `init_db_size = 1GB`, `grow_db_size = 500Mb`, `resize_threshold=100Mb`

hansieodendaal commented:
> One or two nits, but overall happy with how you've done it. I think that RESIZE_THRESHOLD should be moved to the config, with a default value as well as initial size and a new variable for size to grow by. This allows users to control disk usage. I think that RESIZE_THRESHOLD should also be a number in MB and not a percentage. The reason for that is the DB is always growing by a constant size which is `max_block_size * block_time`, so we could use the resize threshold to always add enough space for `x` number of blocks, rather than keeping 50% for growth.
> I suggest `init_db_size = 1GB`, `grow_db_size = 500Mb`, `resize_threshold=100Mb`

@mikethetike and @delta1, I do not think these should be user-configurable, or to change it as suggested, as it would then be prone to errors, overly complicating the code to guard against that. Currently, we have 3 LMDB databases, each with their own size. We could just choose settings that would be small enough to preserve disk space. For example, the blockchain database could be 250MB (currently 1000MB), the peer database and wallet peer database could each be 10 MB (currently 50 MB), all with with RESIZE_THRESHOLD at 50%, increasing by init size.

stringhandler commented:
@hansieodendaal all user configs are prone to error. To be fair, the two peer_db databases don't need to be very configurable, but the main block chain database should be. We can guard against configuration values that don't make sense simply by checking the values provided (e.g. `resize_threshold` > `grow_db_size`) and throw a `ConfigurationError`. One user may decide to allocate the full DB size before syncing, and another may prefer to only grow the db by 1mb at a time, keeping it as small as possible. A user also may not have enough space to grow the db to its full extent (e.g. if we set it to 10gb), but would like to run it for a few more blocks before moving to different hardware.

hansieodendaal commented:
In this case, maybe we can implement a minimum init and grow size to be something like `min_size = 10 x largest block size`
Also `resize_threshold < grow_db_size` and `resize_threshold < min_size`

stringhandler commented:
Personally I like the option of being able to set `grow_size` (or whatever we call it) to `0`, which should error if it ever tries to resize and must be attended to manually. I also think that `resize_threshold` should have no minimum to give the user the most control. They may want to resize on every block. As long as they could not put in a value that would result in an infinite loop, or other crash, I don't think we need to restrict the YOLO users. We should however, choose responsible defaults that don't affect performance but also save on disk space

hansieodendaal commented:
> Personally I like the option of being able to set `grow_size` (or whatever we call it) to `0`, which should error if it ever tries to resize and must be attended to manually. 

Do not think this is wise; will be a user support nightmare. Do not invite what can go wrong. The grow size must be safely larger than the size of any write operation.

> I also think that `resize_threshold` should have no minimum to give the user the most control. 

We must guard against writing a block to the database without enough space left, similar to the above comment, so there must be a minimum value.

delta1 commented:
Personally I think these should be user configurable, with sensible defaults, and guards in place when reading them from config. We can hash out the finer details.

delta1 commented:
The initial 1000 MB gets us to approx 73000 blocks which is about 13 to 14 KB blocks on average. 

With a maximum block size of 1 MB, and 720 blocks per day, it's *possible* that the database might require an extra 720 MB per day (although likely to be lower than that). 

So I agree with @mikethetike for *defaults* - start at 1000 MB, resize when there is 100 MB left, and increase the db size by 500 MB. 

When reading in the configuration values, I suggest to check that the following are true: 
- initial size must be at least 100 MB
- growth size must be at least 10 MB 
- resize threshold (resize with this much space left) must be at least 1 MB, and must be **less than** growth size. ~50% of the initial db size.~ 

@mikethetike @hansieodendaal can we please finalise these figures for implementation?

stringhandler commented:
Agree with those figures, just not sure about the reason for this one: 

> and must be less than 50% of the initial db size.

I think resize threshold should be smaller than growth size otherwise it will constantly be resizing

delta1 commented:
It was just an arbitrarily chosen limit. Happy with your suggestion - will check that resize threshold is less than growth size.

hansieodendaal commented:
Just keep in mind we may very well increase the maximum block size, as the current sizing (maximum block weight) was done for 1 minute target block times, which have doubled to 2 minutes. Therefore, 1 MB is too tight and must include a safety margin. My proposal:

- initial size must be at least 100 MB (_as above_)
- growth size must be at least 20 MB
- resize threshold (resize with this much space left) must be at least 10 MB

delta1 commented:
Good point. Implementing as such.

Commit:
Implement LMDB automatic resizing

This commit implements the functionality to automatically resize LMDB environment mapsize as it fills up.
This should ensure that new data can always be inserted as long as there is disk space available.

There were a few minor issues fixed: spelling, docs, and removing an unused use declaration.

Thanks to @hansieodendaal for his contributions
